crates/sui-json-rpc/src/governance_api.rs

```diff
@@ -35,49 +38,126 @@ impl GovernanceReadApi {
             .get_move_objects(owner, &StakedSui::type_())
             .await?)
     }
-    async fn get_delegations(&self, owner: SuiAddress) -> Result<Vec<Delegation>, Error> {
-        Ok(self
-            .state
-            .get_move_objects(owner, &Delegation::type_())
-            .await?)
+
+    async fn get_delegated_stakes(&self, owner: SuiAddress) -> Result<Vec<DelegatedStake>, Error> {
+        let stakes = self.get_staked_sui(owner).await?;
+        if stakes.is_empty() {
+            return Ok(vec![]);
+        }
+
+        let pools = stakes
+            .into_iter()
+            .fold(BTreeMap::<_, Vec<_>>::new(), |mut pools, s| {
+                pools.entry(s.pool_id()).or_default().push(s);
+                pools
+            });
+
+        let system_state = self.get_system_state()?;
+
+        let mut delegated_stakes = vec![];
+        for (pool_id, stakes) in pools {
+            let validator_address: SuiAddress = self
+                .state
+                .read_table_value(system_state.staking_pool_mappings(), &pool_id)
+                .await
+                .ok_or_else(|| {
+                    Error::UnexpectedError(format!(
+                        "Cannot find validator mapping for staking pool {pool_id}"
+                    ))
+                })?;
+            let pool = self
+                .get_staking_pool(&system_state, &pool_id)
+                .await
+                .ok_or_else(|| {
+                    Error::UnexpectedError(format!("Cannot find staking pool [{pool_id}]"))
+                })?;
+
+            let current_rate: PoolTokenExchangeRate = self
+                .state
+                .read_table_value(&pool.exchange_rates, &system_state.epoch())
+                .await
+                .ok_or_else(|| {
+                    Error::UnexpectedError(format!(
+                        "Cannot find exchange rate for pool [{pool_id}] at epoch {}",
+                        system_state.epoch()
+                    ))
+                })?;
+
+            let mut delegations = vec![];
+            for stake in stakes {
+                // delegation will be active in next epoch
+                let status = if system_state.epoch() >= stake.request_epoch() {
+                    let stake_rate: PoolTokenExchangeRate = self
+                        .state
+                        .read_table_value(&pool.exchange_rates, &stake.request_epoch())
```

Comment:
There might be epochs where we skip exchange rate updates for that epoch because we are in safe mode. I merged a Move PR that dealt with this: https://github.com/MystenLabs/sui/pull/8989. Is it possible to call the `pool_token_exchange_rate_at_epoch` move function from here so we don't have to impl the same logic...

Reply:
I don't think we can call move function from rust? @tnowacki ?